Get test coverage to 100%

jon-betts opened the pull request:
There were a number of problems getting this done:

 * A lot of how this works is bound up in patching `pywb`
 * I can't really think of a non functional way to test most of this
 * `pywb` somehow breaks coverage reports... I don't know how but it's weird
 * The app you start starts another in a separate process, so you can't patch things

So the approach has been to:

 * Unit test what we can
 * Functionally test the rest
 * Expand the coverage report to span both sets of tests
 * Liberally apply `pragma: no cover` to things which clearly should be covered by running the tests
 * Run a small, but real web server in a separate process to have something reliable to call

As an interesting aside, because the functional tests make more extensive use of session scoped fixtures, they are actually much faster than the unit tests. This is because they largely amount to a single call which we now make multiple separate assertions about. 

I wonder if there's scope to make more use of that technique when we have a single object we want to make more assertions about?

jon-betts commented on viahtml/wsgi.py:
```diff
@@ -0,0 +1,11 @@
+"""The application providing a WSGI entry-point."""
+
+from gevent.monkey import patch_all
```
This creates all sorts of horror if it's not called before certain things, which made importing the previous `Application` a dicey business. This doesn't remove the horror, but it does quarantine it

jon-betts commented on viahtml/app.py:
```diff
@@ -23,7 +19,7 @@ def create(cls):
 
         # Move into the correct directory as template paths are relative
-        os.chdir("viahtml")
+        os.chdir(resource_filename("viahtml", "."))
```
Allow this to run from any working directory

jon-betts commented on tox.ini:
```diff
@@ -60,7 +60,7 @@ commands =
     checkformatting: black --check viahtml tests bin
     checkformatting: isort --quiet --check-only viahtml tests bin
-    tests: coverage run -m pytest {posargs:tests/unit/}
+    tests: coverage run -m pytest {posargs:tests/}
```
Run the lot. There's not point in doing it separately at the moment. Also the functional tests are faster in this case

seanh replied:
I think we should change this to be the same as our other apps in some ways:

1. `make test` should run the unit tests only
2. There should be a separate `make functests` command to run the functional tests
3. Only the unit tests should count towards test coverage, not the functional tests

This is in line with our other apps, and coverage isn't a very useful signal with functional tests. I can imagine functional test coverage being difficult to maintain, too. When you have a gap in `viahtml/foo.py`'s unit test coverage it's relatively easy to add another test to `tests/unit/viahtml/foo_test.py` (or to find the existing `foo_test.py` test that should be covering the missed lines but isn't working). But if a change causes the _functional_ tests to have a gap in coverage of a certain source file, I can imagine it being relatively difficult to figure out what the problem is and how to fix it, whether there's a test that supposed to be covering these lines but isn't because it's misbehaving and if so which test and how to fix it, or whether there's a missing test that should be added. It's just a lot harder to know exactly what lines are / should be covered under what test(s). You lack the 1:1 relationship between source module and test module and so much more code is being executed under each test, so much more is going on under each test, that it's much harder to understand.

Of course it's going to be impossible to get 100% test coverage out of viahtml by use of unit tests only, due to the nature of viahtml and its relationship to pywb. I think the solution to this should be the same solution that we apply to our other apps when we have some code that is too difficult or not worthwhile to unit test: use `# pragma: nocover` comments and/or ignore entire files in `.coveragerc`. The difference will be in degree: viahtml will make a lot more use of coverage ignores.

The coverage ignores might function as handy warning signs also: warning developers when they're touching parts of viahtml that don't have unit test coverage.

jon-betts commented on tests/functional/conftest.py:
```diff
@@ -0,0 +1,58 @@
+from viahtml.app import Application
+
+
+@pytest.fixture(scope="session")  # pragma: no cover
```
I've no idea what's up with the coverage, but once you've created the main app all hell breaks lose and nothing makes sense any more. These fixture lines were apparently not covered, but that's total nonsense as they are declarative.

seanh replied:
I'm wondering about the session scope of this and the `with_environ` fixture below. Could this couple tests together in such a way that later tests depend on earlier ones having been run before? In theory both fixtures definitely could do

jon-betts replied:
I don't think so. As this returns an app the tests are making requests to it which are individual. Our app shouldn't store state between requests that could modify the next.

We actually have a single baked request for most of the tests, so this further reduces the chance of any state between requests, as there is only one for the majority. As long as we don't mutate the response in our tests they can't become dependent.

seanh commented on tests/functional/conftest.py:
```diff
@@ -0,0 +1,58 @@
+    return webtest.TestApp(app)
+
+
+@pytest.fixture(scope="session")  # pragma: no cover
+def with_environ():
+    # WSGI uses repeated elements to express this, which means we can't use
+    # the standard configparser.ConfigParser to read them. So they are
+    # duplicated here:
+    os.environ.update(
+        {
+            "VIA_H_EMBED_URL": "http://localhost:3001/hypothesis",
+            "VIA_IGNORE_PREFIXES": "http://localhost:5000/,http://localhost:3001/,https://localhost:5000/,https://localhost:3001/",
+            "VIA_DEBUG": "1",
+        }
+    )
```
I'm not sure what the `with_` in the name is about?

```suggestion
@pytest.fixture(scope="session")  # pragma: no cover
def environ():
    # WSGI uses repeated elements to express this, which means we can't use
    # the standard configparser.ConfigParser to read them. So they are
    # duplicated here:
    os.environ.update(
        {
            "VIA_H_EMBED_URL": "http://localhost:3001/hypothesis",
            "VIA_IGNORE_PREFIXES": "http://localhost:5000/,http://localhost:3001/,https://localhost:5000/,https://localhost:3001/",
            "VIA_DEBUG": "1",
        }
    )
    return os.environ
```

jon-betts replied:
I think you've sort of spotted it and missed it at the same time. When we intend fixtures to be using only to establish a situation rather than to be used as a return value, we do two things to indicate that:

 * Verbify the name
 * Remove the return value

For example a fixture you intend to use the value of might be called 'feature_flag' but if it's only intended to create a situation it might be called 'with_feature_flag_on` and not return anything.

Usually we do `mark.use_fixtures` with these types of fixture, but in this case it didn't work. I could also just inline this into the app fixture and avoid the whole question?

seanh commented on tests/functional/conftest.py:
```diff
@@ -0,0 +1,58 @@
+
+
+@pytest.fixture(scope="session")  # pragma: no cover
+def app(with_environ):
```
```suggestion
@pytest.mark.usefixtures("environ")
def app():
```

Alternatively, I think the `environ` / `with_environ` fixture below could just be `autouse=True`? If `app` depends on it then I think it's getting used in every test. Declare the fact that `environ` is always used in `environ`'s own `@pytest.fixture(...)` instead of indirectly by making `app` depend on it

jon-betts replied:
The environ changes must be in place before the app is created. So we must force the order or it won't start.

seanh commented:
> As an interesting aside, because the functional tests make more extensive use of session scoped fixtures, they are actually much faster than the unit tests. This is because they largely amount to a single call which we now make multiple separate assertions about.
> 
> I wonder if there's scope to make more use of that technique when we have a single object we want to make more assertions about?

On a smaller scale I think it's acceptable to write tests like this:

```python
def test_something():
    # Do some setup.
    ...

    # Call the method under test.
    something()

    # Now assert multiple different things.
    assert something
    assert something_else
    assert another_thing
```

That is: I don't think it's necessary to write a separate test for each assert, or for each thing we want to test. We _should_ write a separate test for each different kind of setup or each different way of calling the method under test (parametrized tests and fixtures can sometimes help to generate the multiple tests here) but it's okay for a single setup and method call to then be followed by multiple unrelated assertions.

On a larger scale we should be very conservative with our use of session-scoped fixtures because they can couple tests together. You can end up with tests that pass when the whole test suite or whole test module is run but fail when only the individual test or test class is run. Or pass and fail depending on what order tests are run in. An innocent change can cause a random test that should be unrelated to start failing. Worse, the same effects can causes tests to pass when they should fail. Basically any across-tests mutable state can cause pretty painful test maintenance issues.

Any fixture that returns an immutable object / doesn't create any mutable state can be session-scoped, although often it's not worth the bother if session-scoping the fixture won't make any measurable performance difference.

For some fixtures that take a long time to run it might be worth finding faster ways to reset the fixture's state inbetween tests, rather than letting pytest re-run the entire fixture every time. Things like app fixtures and DB sessions. But I think the solutions to these aren't always easy, sometimes may not exist.

Limiting the scope of autouse fixtures is another technique here. Minimizing autouse fixtures with a large scope (e.g. in `conftest.py` files), having them only be used by the tests that actually need them instead. Our global services fixtures in LMS are an example of this: there's a single `conftest.py` fixture for each service but they aren't autoused, rather each test module has to declare which fixtures it needs. Technically individual test classes or even test methods could activate service fixtures, but practically it's easier to just activate them on a per-test-module basis with a `pytestmark` (the ease-of-maintenance difference is significant and the performance difference is insignificant)

seanh commented on tests/unit/viahtml/app_test.py:
```diff
@@ -0,0 +1,90 @@
+    def test_it_configures_logging(self, debug_enabled):
+        os.environ["VIA_DEBUG"] = debug_enabled
+
+        with patch("viahtml.app.logging", autospec=True) as logging:
+            Application.create()
+
+            logging.basicConfig.assert_called_once_with(
+                format=Any.string(),
+                level=logging.DEBUG if debug_enabled else logging.INFO,
+            )
```
Should this be a module-level patch fixture? We'd normally do:

```python
@pytest.fixture(autouse=True)
def logging(patch):
    return patch("viahtml.app.logging")
```

Just separates the patching from the test itself and is consistent with how we normally patch things

jon-betts replied:
This is now a separate fixture.

seanh reviewed (changes requested):
Various suggestions, the big ones are:

I don't think we should use session-scoped fixtures when there's no reason to. Some of the session-scoped fixtures seem to function fine, and just as fast, when not session-scoped, and not session-scoping them removes the risk of accidentally coupling tests together. We should limit session-scoped fixtures to ones that won't work if not session scoped or that cause a significant slowdown is not session-scoped.

The coverage ignores can be removed if you enable coverage's gevent support.

I think we should keep things the same as our other apps in terms of `make test` running the unit tests only, a separate `make functests` to run the functests, and only the functests contribute to coverage. We should then use coverage ignores liberally to excuse things that aren't practical to cover with unit tests (which is the same as we do in our other apps, only viahtml is gonna be doing more of it).

Other minor suggestions

seanh commented on tests/unit/viahtml/app_test.py:
```diff
@@ -0,0 +1,90 @@
+
+        assert isinstance(app, FrontEndApp)
+
+    @patch("viahtml.app.apply_pre_app_hooks", autospec=True)
```
We should copy-paste [the `patch` fixture](https://github.com/hypothesis/lms/blob/869911e7093b6cabfca41e9a64828725905f8f38/tests/conftest.py#L62-L74) from our other apps and use that. It's easier to use, integrates with pytest fixtures (class, module or package-level autousing, overriding, ...), ensures that `autospec` is always used, and is consistent with the rest of our apps

jon-betts replied:
We could, and possibly should elsewhere. I also think we should be familiar with the tools available in the libraries we use, rather than using our own re-inventions 100% of the time.

For the reasons you mentioned we should do something like that when we intend to re-use the fixture.

I think we should use this pattern when we need a single fixture once in one place. It's much more concise, and keeps everything relevant near the test in question (no scrolling).

jon-betts commented on tests/unit/viahtml/app_test.py:
```diff
@@ -0,0 +1,90 @@
+from viahtml.app import Application
+
+
+# For reasons I find absolutely mysterious, everytime we create our app
+# coverage can no longer be detected on lines following it. We absolutely get
+# there but for some reason it doesn't show in the report?
+class TestApplicationCreate:  # pragma: no cover
```
Yep, that's it. I had no idea that was a thing. This is done and I've removed the pragma lines.

jon-betts commented on tests/functional/conftest.py:
```diff
@@ -0,0 +1,58 @@
+        yield
+
+
+@pytest.fixture(scope="session")  # pragma: no cover
```
Yep this works fine. It's fast with or without. I've not applied the suggestion as I've removed the pragma in a different commit, but I've done the same thing.

jon-betts commented on tests/simple_server.py:
```diff
@@ -0,0 +1,43 @@
+    :param content_type: Mime type to return for content
+    :param port: Port to serve content on
+    """
+    # A relatively simple WSGI application. It's going to print out the
+    # environment dictionary after being updated by setup_testing_defaults
```
No it's nonsense which originated where I found the example and has piggy-backed along as the code has mutated. I've removed this.

jon-betts commented on tests/unit/conftest.py:
```diff
@@ -0,0 +1,21 @@
+import pytest
+
+
+@pytest.fixture(autouse=True)
+def with_environ():
+    # WSGI uses repeated elements to express this, which means we can't use
+    # the standard configparser.ConfigParser to read them. So they are
+    # duplicated here:
+    os.environ.update(
+        {
+            "VIA_H_EMBED_URL": "http://localhost:3001/hypothesis",
+            "VIA_IGNORE_PREFIXES": (
+                # This is one string
+                "http://localhost:5000/,http://localhost:3001/,"
+                "https://localhost:5000/,https://localhost:3001/"
+            ),
+            "VIA_DEBUG": "1",
+        }
+    )
```
Hmm, not sure where the comment replying to this is, but basically this is a situational fixture. Our pattern for this is to not return the value, and to name it differently.

jon-betts commented on tests/unit/viahtml/app_test.py:
```diff
@@ -0,0 +1,90 @@
+    def test_it_reads_options_from_environment_variables(
+        self, variable, value, expected, Hooks
+    ):
+        os.environ[variable] = value
```
That's done now. This was also useful for the `os.path.exists` test.

I think we can patch it out here, but not in the functional tests, as `pywb` also reads from `os`. There's no sensible place to patch it from what I can tell in that situation. I suppose you could patch it twice?

jon-betts commented:
I've addressed most of the feedback in this PR:

 * The request fixture in the functional tests is no longer "session" scoped
 * Some other fixtures have been added / moved
 * Some comment changes

The remaining issue is to resolve how to do test coverage. This is a longer discussion which I don't think can be fruitfully solved in text. The above has too many things it it to discuss individually. So I'll try and summarise my main thoughts:

 * I suspect there's a way to satisfy the dual need to get actual coverage reporting and still be "like" other products in some aspect which may be enough to satisfy the desire for consistency (e.g. `make testcombined` or something, maintain separate coverage targets for )
 * Not all situations are the same - What is true for one product is not always true for others. The nuance matters.
 * This product cannot achieve 100% line coverage with unit tests
 * 100% line coverage _never_ means 100% test coverage, and it's even more true here than normal
 * Removing automated monitoring of coverage over the functional tests means we remove it from our measurements
 * If we aren't measuring it, we won't maintain it
 * This means our _measured_ quality may increase, but our _real_ test quality will decrease
